File: src/main/java/me/soknight/minigram/chats/service/ChatService.java
package me.soknight.minigram.chats.service;

import lombok.AllArgsConstructor;
import me.soknight.minigram.chats.exception.ApiException;
import me.soknight.minigram.chats.model.attribute.ChatMemberRole;
import me.soknight.minigram.chats.model.attribute.ChatType;
import me.soknight.minigram.chats.model.dto.ChatDto;
import me.soknight.minigram.chats.model.dto.ChatMemberDto;
import me.soknight.minigram.chats.model.dto.mapper.ChatDtoMapper;
import me.soknight.minigram.chats.model.entity.ChatEntity;
import me.soknight.minigram.chats.model.entity.ChatMemberEntity;
import me.soknight.minigram.chats.model.request.CreateChatRequest;
import me.soknight.minigram.chats.model.request.EditChatRequest;
import me.soknight.minigram.chats.model.websocket.ChatEvent;
import me.soknight.minigram.chats.repository.ChatRepository;
import me.soknight.minigram.chats.service.client.ProfileClient;
import me.soknight.minigram.chats.service.client.model.attribute.RelationStatus;
import me.soknight.minigram.chats.service.client.model.attribute.RelationType;
import org.jspecify.annotations.NonNull;
import org.jspecify.annotations.Nullable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.LinkedHashSet;
import java.util.Set;
import java.util.UUID;

@Service
@AllArgsConstructor
public class ChatService {

    private final @NonNull ChatRepository chatRepository;
    private final @NonNull ProfileClient profileClient;
    private final @NonNull ChatEventPublisher eventPublisher;
    private final @NonNull ChatDtoMapper chatDtoMapper;

    @Transactional(readOnly = true)
    public @NonNull Page<ChatDto> getChats(@NonNull Pageable pageable) throws ApiException {
        var actorProfileId = profileClient.resolveMyProfileId();
        var page = chatRepository.findAllByMemberUserId(actorProfileId, pageable);

        try {
            var chats = page.getContent().stream()
                    .map(this::toUncheckedChatDto)
                    .toList();

            return new PageImpl<>(chats, pageable, page.getTotalElements());
        } catch (ChatDtoMapper.ChatDtoMappingException ex) {
            throw ex.apiException();
        }
    }

    @Transactional(readOnly = true)
    public @NonNull ChatDto getChat(long chatId) throws ApiException {
        var actorProfileId = profileClient.resolveMyProfileId();
        return chatDtoMapper.toChatDto(getAccessibleChat(chatId, actorProfileId));
    }

    @Transactional
    public @NonNull ChatDto createChat(@NonNull CreateChatRequest request) throws ApiException {
        var ownerProfileId = profileClient.resolveMyProfileId();
        Set<UUID> memberIds = new LinkedHashSet<>();
        if (request.memberIds() != null)
            memberIds.addAll(request.memberIds());

        memberIds.remove(ownerProfileId);

        validateChatCreation(request.type(), request.title(), memberIds);

        var title = normalizeTitle(request.title(), request.type());
        var chat = new ChatEntity(request.type(), title, ownerProfileId);

        chat.getMembers().add(new ChatMemberEntity(chat, ownerProfileId, ChatMemberRole.OWNER));

        for (UUID memberId : memberIds)
            chat.getMembers().add(new ChatMemberEntity(chat, memberId, ChatMemberRole.MEMBER));

        var dto = chatDtoMapper.toChatDto(chatRepository.save(chat));
        eventPublisher.publish(dto.id(), ChatEvent.chatCreated(dto));
        return dto;
    }

    @Transactional
    public @NonNull ChatDto editChat(long chatId, @NonNull EditChatRequest request) throws ApiException {
        var actorProfileId = profileClient.resolveMyProfileId();
        var chat = getAccessibleChat(chatId, actorProfileId);

        if (!chat.getOwnerId().equals(actorProfileId))
            throw new ApiException(HttpStatus.FORBIDDEN, "access_denied", "Only chat owner can edit the chat");

        if (!chat.isGroup())
            throw new ApiException(HttpStatus.CONFLICT, "chat_edit_not_supported", "Only group chats can be edited");

        if (request.title() != null) {
            var title = normalizeTitle(request.title(), chat.getType());
            chat.updateTitle(title);
        }

        var dto = chatDtoMapper.toChatDto(chat);
        eventPublisher.publish(dto.id(), ChatEvent.chatUpdated(dto));
        return dto;
    }

    @Transactional
    public @NonNull ChatDto deleteChat(long chatId) throws ApiException {
        var actorProfileId = profileClient.resolveMyProfileId();
        var chat = getAccessibleChat(chatId, actorProfileId);

        if (!chat.getOwnerId().equals(actorProfileId))
            throw new ApiException(HttpStatus.FORBIDDEN, "access_denied", "Only chat owner can delete the chat");

        var dto = chatDtoMapper.toChatDto(chat);
        var memberProfileIds = dto.members().stream()
                .map(ChatMemberDto::profileId)
                .toList();

        chatRepository.delete(chat);
        eventPublisher.publishToUsers(memberProfileIds, ChatEvent.chatDeleted(dto));
        return dto;
    }

    @NonNull ChatEntity getAccessibleChat(long chatId, UUID profileId) throws ApiException {
        return chatRepository.findAccessibleById(chatId, profileId)
                .orElseThrow(() -> new ApiException(
                        HttpStatus.NOT_FOUND,
                        "chat_not_found",
                        "Chat {0} not found",
                        chatId
                ));
    }

    private void validateChatCreation(
            @NonNull ChatType type,
            @Nullable String title,
            @NonNull Set<UUID> memberIds
    ) throws ApiException {
        switch (type) {
            case SAVED -> {
                if (!memberIds.isEmpty())
                    throw new ApiException(
                            HttpStatus.BAD_REQUEST,
                            "invalid_chat_members",
                            "Saved messages chat cannot contain other members"
                    );
            }

            case DIRECT -> {
                if (memberIds.size() != 1)
                    throw new ApiException(
                            HttpStatus.BAD_REQUEST,
                            "invalid_chat_members",
                            "Direct chat must contain exactly one additional member"
                    );

                validateFriendRelations(memberIds, "Direct chat can be created only with friends");
            }

            case GROUP -> {
                if (hasBlankTitle(title))
                    throw new ApiException(
                            HttpStatus.BAD_REQUEST,
                            "invalid_chat_title",
                            "Group chat title must not be blank"
                    );

                validateFriendRelations(memberIds, "Only friends can be added to a group chat");
            }
        }
    }

    private void validateFriendRelations(@NonNull Set<UUID> memberIds, @NonNull String message) throws ApiException {
        for (UUID memberId : memberIds) {
            var relation = profileClient.getRelation(memberId, RelationType.OUTGOING);
            var status = relation.status();

            if (status == RelationStatus.FRIEND) continue;

            throw new ApiException(
                    HttpStatus.FORBIDDEN,
                    "relation_not_accepted",
                    "{0} Relation status with user {1} is {2}",
                    message,
                    memberId,
                    status == null ? "null" : status.getKey()
            );
        }
    }

    private @Nullable String normalizeTitle(@Nullable String rawTitle, @NonNull ChatType type) {
        if (type != ChatType.GROUP) return null;
        return hasBlankTitle(rawTitle) ? null : rawTitle.trim();
    }

    private boolean hasBlankTitle(@Nullable String value) {
        return value == null || value.isBlank();
    }

    private @NonNull ChatDto toUncheckedChatDto(@NonNull ChatEntity chat) {
        try {
            return chatDtoMapper.toChatDto(chat);
        } catch (ApiException ex) {
            throw new ChatDtoMapper.ChatDtoMappingException(ex);
        }
    }

}

File: src/main/kotlin/me/soknight/minigram/chats/service/ChatService.kt
@file:OptIn(ExperimentalUuidApi::class)

package me.soknight.minigram.chats.service

import kotlinx.serialization.encodeToString
import me.soknight.minigram.chats.client.ProfileClient
import me.soknight.minigram.chats.dto.ChatDto
import me.soknight.minigram.chats.dto.request.CreateChatRequest
import me.soknight.minigram.chats.dto.request.EditChatRequest
import me.soknight.minigram.chats.events.ChatEvent
import me.soknight.minigram.chats.exception.*
import me.soknight.minigram.chats.model.ChatMemberRole
import me.soknight.minigram.chats.model.ChatType
import me.soknight.minigram.chats.model.RelationStatus
import me.soknight.minigram.chats.model.RelationType
import me.soknight.minigram.chats.plugin.appJson
import me.soknight.minigram.chats.repository.ChatMemberRepository
import me.soknight.minigram.chats.repository.ChatRepository
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid
import org.koin.core.annotation.Single

@Single
class ChatService(
    private val chatRepository: ChatRepository,
    private val memberRepository: ChatMemberRepository,
    private val dtoMapper: ChatDtoMapper,
    private val eventPublisher: ChatEventPublisher
) {
    suspend fun getChats(userId: Uuid, profileClient: ProfileClient, page: Int, size: Int): List<ChatDto> =
        chatRepository.findAllByMemberId(userId, page, size).map { dtoMapper.toChatDto(it, profileClient) }

    suspend fun getChat(chatId: Long, userId: Uuid, profileClient: ProfileClient): ChatDto {
        val chat = chatRepository.findAccessibleById(chatId, userId) ?: throw ChatNotFoundException(chatId)
        return dtoMapper.toChatDto(chat, profileClient)
    }

    suspend fun createChat(request: CreateChatRequest, userId: Uuid, profileClient: ProfileClient): ChatDto {
        when (request.type) {
            ChatType.SAVED -> {
                if (request.memberIds.isNotEmpty())
                    throw InvalidChatMembersException("SAVED chats cannot have other members")
            }
            ChatType.DIRECT -> {
                if (request.memberIds.size != 1)
                    throw InvalidChatMembersException("DIRECT chats require exactly 1 other member")
                assertFriend(request.memberIds[0], profileClient)
            }
            ChatType.GROUP -> {
                if (request.memberIds.isEmpty())
                    throw InvalidChatMembersException("GROUP chats require at least 1 other member")
                if (request.title.isNullOrBlank()) throw InvalidChatTitleException()
                request.memberIds.forEach { assertFriend(it, profileClient) }
            }
        }

        val chat = chatRepository.insert(request.type, request.title, userId)
        memberRepository.insert(chat.id, userId, ChatMemberRole.OWNER)
        request.memberIds.forEach { memberRepository.insert(chat.id, it, ChatMemberRole.MEMBER) }

        val dto        = dtoMapper.toChatDto(chat, profileClient)
        val allMembers = listOf(userId) + request.memberIds
        eventPublisher.publishToUsers(allMembers, ChatEvent(
            ChatEvent.Type.CHAT_CREATED, chat.id,
            appJson.encodeToJsonElement(ChatDto.serializer(), dto)
        ))
        return dto
    }

    suspend fun editChat(chatId: Long, request: EditChatRequest, userId: Uuid, profileClient: ProfileClient): ChatDto {
        val chat = chatRepository.findAccessibleById(chatId, userId) ?: throw ChatNotFoundException(chatId)
        if (chat.ownerId != userId) throw AccessDeniedException()
        if (chat.type != ChatType.GROUP) throw ChatEditNotSupportedException()
        chatRepository.updateTitle(chatId, request.title)
        val updated = chatRepository.findById(chatId)!!
        val dto     = dtoMapper.toChatDto(updated, profileClient)
        eventPublisher.publish(chatId, ChatEvent(
            ChatEvent.Type.CHAT_UPDATED, chatId,
            appJson.encodeToJsonElement(ChatDto.serializer(), dto)
        ))
        return dto
    }

    suspend fun deleteChat(chatId: Long, userId: Uuid) {
        val chat    = chatRepository.findAccessibleById(chatId, userId) ?: throw ChatNotFoundException(chatId)
        if (chat.ownerId != userId) throw AccessDeniedException()
        val members = memberRepository.findUserIdsByChatId(chatId)
        eventPublisher.publishToUsers(members, ChatEvent(ChatEvent.Type.CHAT_DELETED, chatId))
        chatRepository.delete(chatId)
    }

    private suspend fun assertFriend(targetId: Uuid, profileClient: ProfileClient) {
        val relation = profileClient.getRelation(targetId, RelationType.OUTGOING)
        if (relation?.status != RelationStatus.FRIEND) throw RelationNotAcceptedException()
    }
}
